File: trunk/EneterProtoBufSerializer/EneterProtoBufSerializer/ProtoBuf/ProtoBufSerializer.cs
/*
 * Project: Eneter.ProtoBuf.Serializer
 * Author:  Ondrej Uzovic
 * 
 * Copyright © Ondrej Uzovic 2013
*/

using System;
using System.IO;
using Eneter.Messaging.DataProcessing.Serializing;
using Eneter.Messaging.EndPoints.Rpc;
using Eneter.Messaging.EndPoints.TypedMessages;
using Eneter.Messaging.MessagingSystems.Composites.MonitoredMessagingComposit;
using Eneter.Messaging.Nodes.Broker;
using Eneter.Messaging.Nodes.ChannelWrapper;
using ProtoBuf;
using Eneter.Messaging.MessagingSystems.Composites.MessageBus;

namespace Eneter.ProtoBuf
{
    /// <summary>
    /// Implements protocol buffer serialization for Eneter.Messaging.Framework.
    /// </summary>
    public class ProtoBufSerializer : ISerializer
    {
        /// <summary>
        /// Serializes data using Protocol Buffer binary format.
        /// </summary>
        /// <typeparam name="T">data type of serialized data</typeparam>
        /// <param name="dataToSerialize">protocol buffer generated data class to be serialized.</param>
        /// <returns>array containing serialized data</returns>
        public object Serialize<T>(T dataToSerialize)
        {
            object aSerializedData;

            // If it is an internal Eneter type adapt it for ProtoBuf
            if (dataToSerialize is MultiTypedMessage)
            {
                aSerializedData = SerializeMultiTypedMessage(dataToSerialize as MultiTypedMessage);
            }
            else if (dataToSerialize is MessageBusMessage)
            {
                aSerializedData = SerializeMessageBusMessage(dataToSerialize as MessageBusMessage);
            }
            else if (dataToSerialize is RpcMessage)
            {
                aSerializedData = SerializeRpcMessage(dataToSerialize as RpcMessage);
            }
            else if (dataToSerialize is EventArgs)
            {
                aSerializedData = SerializeEventArgs(dataToSerialize as EventArgs);
            }
            else if (dataToSerialize is WrappedData)
            {
                aSerializedData = SerializeWrappedData(dataToSerialize as WrappedData);
            }
            else if (dataToSerialize is BrokerMessage)
            {
                aSerializedData = SerializeBrokerMessage(dataToSerialize as BrokerMessage);
            }
            else if (dataToSerialize is MonitorChannelMessage)
            {
                aSerializedData = SerializeMonitorChannelMessage(dataToSerialize as MonitorChannelMessage);
            }
            else if (dataToSerialize is VoidMessage)
            {
                aSerializedData = SerializeVoidMessage(dataToSerialize as VoidMessage);
            }
            else
            {
                // If it is not internal Eneter message use directly ProtoBuf serializer.
                aSerializedData = SerializeProtoBuf<T>(dataToSerialize);
            }

            return aSerializedData;
        }

        /// <summary>
        /// Deserializes data from the Protocol Buffer binary format.
        /// </summary>
        /// <typeparam name="T">data type of deserialized data.</typeparam>
        /// <param name="serializedData">data (byte[]) serialized by protocol buffer to be deserialized</param>
        /// <returns>instance of deserialized data type</returns>
        public T Deserialize<T>(object serializedData)
        {
            using (MemoryStream aBuf = new MemoryStream((byte[])serializedData))
            {
                T aDeserializedObject;

                // If it is an internal Eneter type, adapt it from the proto type.
                if (typeof(T) == typeof(MultiTypedMessage))
                {
                    aDeserializedObject = (T)((object)DeserializeMultiTypedMessage(aBuf));
                }
                else if (typeof(T) == typeof(MessageBusMessage))
                {
                    aDeserializedObject = (T)((object)DeserializeMessageBusMessage(aBuf));
                }
                else if (typeof(T) == typeof(RpcMessage))
                {
                    aDeserializedObject = (T)((object)DeserializeRpcMessage(aBuf));
                }
                else if (typeof(T) == typeof(EventArgs))
                {
                    aDeserializedObject = (T)((object)DeserializeEventArgs(aBuf));
                }
                else if (typeof(T) == typeof(WrappedData))
                {
                    aDeserializedObject = (T)((object)DeserializeWrappedData(aBuf));
                }
                else if (typeof(T) == typeof(BrokerMessage))
                {
                    aDeserializedObject = (T)((object)DeserializeBrokerMessage(aBuf));
                }
                else if (typeof(T) == typeof(MonitorChannelMessage))
                {
                    aDeserializedObject = (T)((object)DeserializeMonitorChannelMessage(aBuf));
                }
                else if (typeof(T) == typeof(VoidMessage))
                {
                    aDeserializedObject = (T)((object)DeserializeVoidMessage(aBuf));
                }
                else
                {
                    // If it is not internal Eneter message use directly ProtoBuf serializer.
                    aDeserializedObject = Serializer.Deserialize<T>(aBuf);
                }

                return aDeserializedObject;
            }
        }

        private byte[] SerializeMultiTypedMessage(MultiTypedMessage data)
        {
            MultiTypedMessageProto aMultiTypedMessageProto = new MultiTypedMessageProto();
            aMultiTypedMessageProto.TypeName = data.TypeName;
            if (data.MessageData != null)
            {
                if (data.MessageData is string)
                {
                    aMultiTypedMessageProto.MessageDataStr = (string)data.MessageData;
                }
                else
                {
                    aMultiTypedMessageProto.MessageDataBin = (byte[])data.MessageData;
                }
            }
            return SerializeProtoBuf<MultiTypedMessageProto>(aMultiTypedMessageProto);
        }

        private MultiTypedMessage DeserializeMultiTypedMessage(MemoryStream data)
        {
            MultiTypedMessageProto aMultiTypedMessageProto = Serializer.Deserialize<MultiTypedMessageProto>(data);
            MultiTypedMessage aMultiTypedMessage = new MultiTypedMessage();
            aMultiTypedMessage.TypeName = aMultiTypedMessageProto.TypeName;
            if (aMultiTypedMessageProto.MessageDataStrSpecified)
            {
                aMultiTypedMessage.MessageData = aMultiTypedMessageProto.MessageDataStr;
            }
            else if (aMultiTypedMessageProto.MessageDataBinSpecified)
            {
                aMultiTypedMessage.MessageData = aMultiTypedMessageProto.MessageDataBin;
            }

            return aMultiTypedMessage;
        }

        private byte[] SerializeMessageBusMessage(MessageBusMessage data)
        {
            MessageBusMessageProto aMessageBusMessageProto = new MessageBusMessageProto();
            aMessageBusMessageProto.Request = (int)data.Request;
            aMessageBusMessageProto.Id = data.Id;
            if (data.MessageData != null)
            {
                if (data.MessageData is string)
                {
                    aMessageBusMessageProto.MessageDataStr = (string)data.MessageData;
                }
                else
                {
                    aMessageBusMessageProto.MessageDataBin = (byte[])data.MessageData;
                }
            }
            return SerializeProtoBuf<MessageBusMessageProto>(aMessageBusMessageProto);
        }

        private MessageBusMessage DeserializeMessageBusMessage(MemoryStream data)
        {
            MessageBusMessageProto aMessageBusMessageProto = Serializer.Deserialize<MessageBusMessageProto>(data);
            MessageBusMessage aMessageBusMessage = new MessageBusMessage();
            aMessageBusMessage.Request = (EMessageBusRequest)aMessageBusMessageProto.Request;
            aMessageBusMessage.Id = aMessageBusMessageProto.Id;
            if (aMessageBusMessageProto.MessageDataStrSpecified)
            {
                aMessageBusMessage.MessageData = aMessageBusMessageProto.MessageDataStr;
            }
            else if (aMessageBusMessageProto.MessageDataBinSpecified)
            {
                aMessageBusMessage.MessageData = aMessageBusMessageProto.MessageDataBin;
            }
            return aMessageBusMessage;
        }

        private byte[] SerializeRpcMessage(RpcMessage data)
        {
            RpcMessageProto anRpcMessageProto = new RpcMessageProto();
            anRpcMessageProto.Id = data.Id;
            anRpcMessageProto.Flag = data.Flag;
            anRpcMessageProto.OperationName = data.OperationName;
            if (data.SerializedData != null)
            {
                foreach (byte[] aMethodParameter in data.SerializedData)
                {
                    anRpcMessageProto.SerializedData.Add(aMethodParameter);
                }
            }
            anRpcMessageProto.ErrorType = data.ErrorType;
            anRpcMessageProto.ErrorMessage = data.ErrorMessage;
            anRpcMessageProto.ErrorDetails = data.ErrorDetails;

            return SerializeProtoBuf<RpcMessageProto>(anRpcMessageProto);
        }

        private RpcMessage DeserializeRpcMessage(MemoryStream data)
        {
            RpcMessageProto anRpcMessageProto = Serializer.Deserialize<RpcMessageProto>(data);
            RpcMessage anRpcMessage = new RpcMessage();
            anRpcMessage.Id = anRpcMessageProto.Id;
            anRpcMessage.Flag = anRpcMessageProto.Flag;
            anRpcMessage.OperationName = anRpcMessageProto.OperationName;
            anRpcMessage.ErrorType = anRpcMessageProto.ErrorType;
            anRpcMessage.ErrorMessage = anRpcMessageProto.ErrorMessage;
            anRpcMessage.ErrorDetails = anRpcMessageProto.ErrorDetails;
            
            if (anRpcMessageProto.SerializedData != null)
            {
                anRpcMessage.SerializedData = new object[anRpcMessageProto.SerializedData.Count];
                for (int i = 0; i < anRpcMessageProto.SerializedData.Count; ++i)
                {
                    anRpcMessage.SerializedData[i] = anRpcMessageProto.SerializedData[i];
                }
            }

            return anRpcMessage;
        }

        private byte[] SerializeEventArgs(EventArgs data)
        {
            return myEmptyMessage;
        }

        private EventArgs DeserializeEventArgs(MemoryStream data)
        {
            Serializer.Deserialize<EventArgsProto>(data);
            return myEventArgs;
        }

        private byte[] SerializeWrappedData(WrappedData data)
        {
            WrappedDataProto aWrappedDataProto = new WrappedDataProto();
            aWrappedDataProto.AddedData = (string)data.AddedData;
            if (data.OriginalData != null)
            {
                if (data.OriginalData is string)
                {
                    aWrappedDataProto.OriginalDataStr = (string)data.OriginalData;
                }
                else
                {
                    aWrappedDataProto.OriginalDataBin = (byte[])data.OriginalData;
                }
            }
            return SerializeProtoBuf<WrappedDataProto>(aWrappedDataProto);
        }

        private WrappedData DeserializeWrappedData(MemoryStream data)
        {
            WrappedDataProto aWrappedDataProto = Serializer.Deserialize<WrappedDataProto>(data);
            WrappedData aWrappedData = new WrappedData();
            aWrappedData.AddedData = aWrappedDataProto.AddedData;
            if (aWrappedDataProto.OriginalDataStrSpecified)
            {
                aWrappedData.OriginalData = aWrappedDataProto.OriginalDataStr;
            }
            else if (aWrappedDataProto.OriginalDataBinSpecified)
            {
                aWrappedData.OriginalData = aWrappedDataProto.OriginalDataBin;
            }

            return aWrappedData;
        }


        private byte[] SerializeBrokerMessage(BrokerMessage data)
        {
            BrokerMessageProto aBrokerMessageProto = new BrokerMessageProto();
            aBrokerMessageProto.Request = (int)data.Request;
            aBrokerMessageProto.MessageTypes.AddRange(data.MessageTypes);
            if (data.Message is string)
            {
                aBrokerMessageProto.MessageStr = (string)data.Message;
            }
            else
            {
                aBrokerMessageProto.MessageBin = (byte[])data.Message;
            }
            return SerializeProtoBuf<BrokerMessageProto>(aBrokerMessageProto);
        }

        private BrokerMessage DeserializeBrokerMessage(MemoryStream data)
        {
            BrokerMessageProto aBrokerMessageProto = Serializer.Deserialize<BrokerMessageProto>(data);
            BrokerMessage aBrokerMessage = new BrokerMessage();
            aBrokerMessage.Request = (EBrokerRequest)aBrokerMessageProto.Request;
            aBrokerMessage.MessageTypes = aBrokerMessageProto.MessageTypes.ToArray();
            if (aBrokerMessageProto.MessageStrSpecified)
            {
                aBrokerMessage.Message = aBrokerMessageProto.MessageStr;
            }
            else if (aBrokerMessageProto.MessageBinSpecified)
            {
                aBrokerMessage.Message = aBrokerMessageProto.MessageBin;
            }
            return aBrokerMessage;
        }


        private byte[] SerializeMonitorChannelMessage(MonitorChannelMessage data)
        {
            MonitorChannelMessageProto aMonitorChannelMessageProto = new MonitorChannelMessageProto();
            aMonitorChannelMessageProto.MessageType = (int)data.MessageType;
            if (data.MessageContent is string)
            {
                aMonitorChannelMessageProto.MessageContentStr = (string)data.MessageContent;
            }
            else
            {
                aMonitorChannelMessageProto.MessageContentBin = (byte[])data.MessageContent;
            }
            return SerializeProtoBuf<MonitorChannelMessageProto>(aMonitorChannelMessageProto);
        }

        private MonitorChannelMessage DeserializeMonitorChannelMessage(MemoryStream data)
        {
            MonitorChannelMessageProto aMonitorChannelMessageProto = Serializer.Deserialize<MonitorChannelMessageProto>(data);
            MonitorChannelMessage aMonitorChannelMessage = new MonitorChannelMessage();
            aMonitorChannelMessage.MessageType = (MonitorChannelMessageType)aMonitorChannelMessageProto.MessageType;
            if (aMonitorChannelMessageProto.MessageContentStrSpecified)
            {
                aMonitorChannelMessage.MessageContent = aMonitorChannelMessageProto.MessageContentStr;
            }
            else if (aMonitorChannelMessageProto.MessageContentBinSpecified)
            {
                aMonitorChannelMessage.MessageContent = aMonitorChannelMessageProto.MessageContentBin;
            }
            return aMonitorChannelMessage;
        }

        private byte[] SerializeVoidMessage(VoidMessage data)
        {
            return myEmptyMessage;
        }

        private VoidMessage DeserializeVoidMessage(MemoryStream data)
        {
            // Read bytes from the stream.
            Serializer.Deserialize<VoidMessageProto>(data);
            return myVoidMessage;
        }

        private byte[] SerializeProtoBuf<_T>(_T dataToSerialize)
        {
            using (MemoryStream aBuf = new MemoryStream())
            {
                Serializer.Serialize<_T>(aBuf, dataToSerialize);
                return aBuf.ToArray();
            }
        }

        private readonly EventArgs myEventArgs = new EventArgs();
        private readonly VoidMessage myVoidMessage = new VoidMessage();
        private readonly byte[] myEmptyMessage = { 8, 0 };
    }
}

File: trunk/eneter-protobuf-android/src/eneter/protobuf/ProtoBufSerializer.java
/**
 * Project: Eneter.ProtoBuf.Serializer
 * Author: Ondrej Uzovic
 * 
 * Copyright © 2013 Ondrej Uzovic
 * 
 */

package eneter.protobuf;

import java.lang.reflect.Method;

import com.google.protobuf.GeneratedMessage;

import eneter.messaging.dataprocessing.serializing.ISerializer;
import eneter.messaging.endpoints.rpc.RpcMessage;
import eneter.messaging.endpoints.typedmessages.VoidMessage;
import eneter.messaging.endpoints.typedmessages.internal.ReliableMessage;
import eneter.messaging.messagingsystems.composites.monitoredmessagingcomposit.*;
import eneter.messaging.nodes.broker.*;
import eneter.messaging.nodes.channelwrapper.WrappedData;


/**
 * Implements protocol buffer serialization for Eneter.Messaging.Framework.
 * 
 *
 */
public class ProtoBufSerializer implements ISerializer
{

    /**
     * Serializes data using Protocol Buffer binary format.
     * @param dataToSerialize protocol buffer generated data class to be serialized.
     * @param clazz data type of serialized data.
     * @return byte[] array containing serialized data. 
     */
    @Override
    public <T> Object serialize(T dataToSerialize, Class<T> clazz)
            throws Exception
    {
        boolean aUseDirectlyProtobBuf = false;
        Object aSerializedData = null;
        
        if (clazz == String.class)
        {
            aSerializedData = PrimitiveTypesWrapper.serializeString((String)dataToSerialize);
        }
        // If it is a primitive type
        else if (clazz.isPrimitive())
        {
            if (clazz == Boolean.class || clazz == Byte.class || clazz == Character.class ||
                clazz == Short.class || clazz == Integer.class || clazz == Long.class ||
                clazz == Float.class || clazz == Double.class)
            {
                throw new IllegalStateException("ProtoBufSerializer does not support primitive types as references (e.g. int is supported but not Integer).");
            }
            
            if (clazz == boolean.class)
            {
                aSerializedData = PrimitiveTypesWrapper.serializeBool((Boolean)dataToSerialize);
            }
            else if (clazz == byte.class)
            {
                aSerializedData = PrimitiveTypesWrapper.serializeByte((Byte)dataToSerialize);
            }
            else if (clazz == char.class)
            {
                aSerializedData = PrimitiveTypesWrapper.serializeChar((Character)dataToSerialize);
            }
            else if (clazz == short.class)
            {
                aSerializedData = PrimitiveTypesWrapper.serializeShort((Short)dataToSerialize);
            }
            else if (clazz == int.class)
            {
                aSerializedData = PrimitiveTypesWrapper.serializeInt((Integer)dataToSerialize);
            }
            else if (clazz == long.class)
            {
                aSerializedData = PrimitiveTypesWrapper.serializeLong((Long)dataToSerialize);
            }
            else if (clazz == float.class)
            {
                aSerializedData = PrimitiveTypesWrapper.serializeFloat((Float)dataToSerialize);
            }
            else if (clazz == double.class)
            {
                aSerializedData = PrimitiveTypesWrapper.serializeDouble((Double)dataToSerialize);
            }
            else
            {
                aUseDirectlyProtobBuf = true;
            }
        }
        // if it is an array of string or a primitive type.
        else if (clazz.isArray())
        {
            if (clazz == Boolean[].class || clazz == Byte[].class || clazz == Character[].class ||
                clazz == Short[].class || clazz == Integer[].class || clazz == Long[].class ||
                clazz == Float[].class || clazz == Double[].class)
            {
                throw new IllegalStateException("ProtoBufSerializer does not support arrays of primitive types references (e.g. int[] is supported but not Integer[]).");
            }
            
            if (clazz == String[].class)
            {
                aSerializedData = PrimitiveTypesWrapper.serializeStringArray((String[])dataToSerialize);
            }
            else if (clazz == boolean[].class)
            {
                aSerializedData = PrimitiveTypesWrapper.serializeBoolArray((boolean[])dataToSerialize);
            }
            else if (clazz == byte[].class)
            {
                aSerializedData = PrimitiveTypesWrapper.serializeByteArray((byte[])dataToSerialize);
            }
            else if (clazz == char[].class)
            {
                aSerializedData = PrimitiveTypesWrapper.serializeCharArray((char[])dataToSerialize);
            }
            else if (clazz == short[].class)
            {
                aSerializedData = PrimitiveTypesWrapper.serializeShortArray((short[])dataToSerialize);
            }
            else if (clazz == int[].class)
            {
                aSerializedData = PrimitiveTypesWrapper.serializeIntArray((int[])dataToSerialize);
            }
            else if (clazz == long[].class)
            {
                aSerializedData = PrimitiveTypesWrapper.serializeLongArray((long[])dataToSerialize);
            }
            else if (clazz == float[].class)
            {
                aSerializedData = PrimitiveTypesWrapper.serializeFloatArray((float[])dataToSerialize);
            }
            else if (clazz == double[].class)
            {
                aSerializedData = PrimitiveTypesWrapper.serializeDoubleArray((double[])dataToSerialize);
            }
            else
            {
                aUseDirectlyProtobBuf = true;
            }
        }
        // If it is an internal Eneter type adapt it for ProtoBuf
        else if (clazz == RpcMessage.class)
        {
            aSerializedData = EneterTypesWrapper.serializeRpcMessage((RpcMessage)dataToSerialize);
        }
        else if (clazz == WrappedData.class)
        {
            aSerializedData = EneterTypesWrapper.serializeWrappedData((WrappedData)dataToSerialize);
        }
        else if (clazz == BrokerMessage.class)
        {
            aSerializedData = EneterTypesWrapper.serializeBrokerMessage((BrokerMessage)dataToSerialize);
        }
        else if (clazz == ReliableMessage.class)
        {
            aSerializedData = EneterTypesWrapper.serializeReliableMessage((ReliableMessage)dataToSerialize);
        }
        else if (clazz == MonitorChannelMessage.class)
        {
            aSerializedData = EneterTypesWrapper.serializeMonitorChannelMessage((MonitorChannelMessage)dataToSerialize);
        }
        else if (clazz == VoidMessage.class)
        {
            aSerializedData = EneterTypesWrapper.serializeVoidMessage((VoidMessage)dataToSerialize);
        }
        else
        {
            aUseDirectlyProtobBuf = true;
        }

        // If it is not string, primitive type or array of an primitive type then
        // try to use directly ProtoBuf serializer.
        if (aUseDirectlyProtobBuf)
        {
            GeneratedMessage aDataToSerialize = (GeneratedMessage)dataToSerialize;
            aSerializedData = aDataToSerialize.toByteArray();
        }
        
        return aSerializedData;
    }

    /**
     * Deserializes data from the Protocol Buffer binary format.
     * @param serializedData data (byte[]) serialized by protocol buffer to be deserialized
     * @param clazz data type of deserialized data.
     * @return instance of deserialized data type 
     */
    @SuppressWarnings("unchecked")
    @Override
    public <T> T deserialize(Object serializedData, Class<T> clazz)
            throws Exception
    {
        boolean aUseDirectlyProtoBuf = false;
        T aDeserializedObject = null;

        if (clazz == String.class)
        {
            aDeserializedObject = clazz.cast(PrimitiveTypesWrapper.deserializeString((byte[])serializedData));
        }
        // If it is a primitive type
        else if (clazz.isPrimitive())
        {
            if (clazz == Boolean.class || clazz == Byte.class || clazz == Character.class ||
                clazz == Short.class || clazz == Integer.class || clazz == Long.class ||
                clazz == Float.class || clazz == Double.class)
            {
                throw new IllegalStateException("ProtoBufSerializer does not support primitive types as references (e.g. int is supported but not Integer).");
            }
            
            if (clazz == boolean.class)
            {
                Boolean aResult = PrimitiveTypesWrapper.deserializeBool((byte[])serializedData);
                aDeserializedObject = (T) aResult;
            }
            else if (clazz == byte.class)
            {
                Byte aResult = PrimitiveTypesWrapper.deserializeByte((byte[])serializedData);
                aDeserializedObject = (T) aResult;
            }
            else if (clazz == char.class)
            {
                Character aResult = PrimitiveTypesWrapper.deserializeChar((byte[])serializedData);
                aDeserializedObject = (T) aResult;
            }
            else if (clazz == short.class)
            {
                Short aResult = PrimitiveTypesWrapper.deserializeShort((byte[])serializedData);
                aDeserializedObject = (T) aResult;
            }
            else if (clazz == int.class)
            {
                Integer aResult = PrimitiveTypesWrapper.deserializeInt((byte[])serializedData);
                aDeserializedObject = (T) aResult;
            }
            else if (clazz == long.class)
            {
                Long aResult = PrimitiveTypesWrapper.deserializeLong((byte[])serializedData);
                aDeserializedObject = (T) aResult;
            }
            else if (clazz == float.class)
            {
                Float aResult = PrimitiveTypesWrapper.deserializeFloat((byte[])serializedData);
                aDeserializedObject = (T) aResult;
            }
            else if (clazz == double.class)
            {
                Double aResult = PrimitiveTypesWrapper.deserializeDouble((byte[])serializedData);
                aDeserializedObject = (T) aResult;
            }
            else
            {
                aUseDirectlyProtoBuf = true;
            }
        }
        // If is an array of string or a primitive type.
        else if (clazz.isArray())
        {
            if (clazz == Boolean[].class || clazz == Byte[].class || clazz == Character[].class ||
                clazz == Short[].class || clazz == Integer[].class || clazz == Long[].class ||
                clazz == Float[].class || clazz == Double[].class)
            {
                throw new IllegalStateException("ProtoBufSerializer does not support arrays of primitive types references (e.g. int[] is supported but not Integer[]).");
            }
            
            if (clazz == String[].class)
            {
                aDeserializedObject = clazz.cast(PrimitiveTypesWrapper.deserializeStringArray((byte[])serializedData));
            }
            else if (clazz == boolean[].class)
            {
                aDeserializedObject = clazz.cast(PrimitiveTypesWrapper.deserializeBoolArray((byte[])serializedData));
            }
            else if (clazz == byte[].class)
            {
                aDeserializedObject = clazz.cast(PrimitiveTypesWrapper.deserializeByteArray((byte[])serializedData));
            }
            else if (clazz == char[].class)
            {
                aDeserializedObject = clazz.cast(PrimitiveTypesWrapper.deserializeCharArray((byte[])serializedData));
            }
            else if (clazz == short[].class)
            {
                aDeserializedObject = clazz.cast(PrimitiveTypesWrapper.deserializeShortArray((byte[])serializedData));
            }
            else if (clazz == int[].class)
            {
                aDeserializedObject = clazz.cast(PrimitiveTypesWrapper.deserializeIntArray((byte[])serializedData));
            }
            else if (clazz == long[].class)
            {
                aDeserializedObject = clazz.cast(PrimitiveTypesWrapper.deserializeLongArray((byte[])serializedData));
            }
            else if (clazz == float[].class)
            {
                aDeserializedObject = clazz.cast(PrimitiveTypesWrapper.deserializeFloatArray((byte[])serializedData));
            }
            else if (clazz == double[].class)
            {
                aDeserializedObject = clazz.cast(PrimitiveTypesWrapper.deserializeDoubleArray((byte[])serializedData));
            }
            else
            {
                aUseDirectlyProtoBuf = true;
            }
        }
        // If it is an internal Eneter type, adapt it from the proto type.
        else if (clazz == RpcMessage.class)
        {
            aDeserializedObject = clazz.cast(EneterTypesWrapper.deserializeRpcMessage((byte[])serializedData));
        }
        else if (clazz == WrappedData.class)
        {
            aDeserializedObject = clazz.cast(EneterTypesWrapper.deserializeWrappedData((byte[])serializedData));
        }
        else if (clazz == BrokerMessage.class)
        {
            aDeserializedObject = clazz.cast(EneterTypesWrapper.deserializeBrokerMessage((byte[])serializedData)); 
        }
        else if (clazz == ReliableMessage.class)
        {
            aDeserializedObject = clazz.cast(EneterTypesWrapper.deserializeReliableMessage((byte[])serializedData)); 
        }
        else if (clazz == MonitorChannelMessage.class)
        {
            aDeserializedObject = clazz.cast(EneterTypesWrapper.deserializeMonitorChannelMessage((byte[])serializedData)); 
        }
        else if (clazz == VoidMessage.class)
        {
            aDeserializedObject = clazz.cast(EneterTypesWrapper.deserializeVoidMessage((byte[])serializedData));
        }
        else
        {
            aUseDirectlyProtoBuf = true;
        }
        
        if (aUseDirectlyProtoBuf)
        {
            // If it is not an Eneter internal type then use directly protobuf.
            Method aParsingMethod = clazz.getMethod("parseFrom", byte[].class);
            aDeserializedObject = clazz.cast(aParsingMethod.invoke(null, (byte[])serializedData));
        }
        
        return aDeserializedObject;
    }
    
    
}
